Add setting to toggle the yank register using the system clipboard

blefevre opened the pull request:
Adds a setting to enable/disable the default yank register (`"`) from mapping to the system clipboard.
Defaults to enabled so there will be no change for existing users.

71 commented on src/state/registers.ts:
```diff
@@ -5,6 +5,7 @@ import { Context, prompt, SelectionBehavior, Selections } from "../api";
 import { ArgumentError, assert, EditNotAppliedError, EditorRequiredError } from "../utils/errors";
 import { noUndoStops } from "../utils/misc";
 import type * as TrackedSelection from "../utils/tracked-selection";
+import { extensionName } from "../utils/constants";
```
nit: please sort imports by import path (not obvious in the GitHub diff):

```suggestion
import { extensionName } from "../utils/constants";
import { ArgumentError, assert, EditNotAppliedError, EditorRequiredError } from "../utils/errors";
import { noUndoStops } from "../utils/misc";
import type * as TrackedSelection from "../utils/tracked-selection";
```

71 commented on package.build.ts:
```diff
@@ -569,6 +569,12 @@ export const pkg = (modules: Builder.ParsedModule[]) => ({
+        "dance.useSystemClipboard": {
+          type: "boolean",
+          default: true,
+          description: "Allow default yank register (\") to map to system clipboard",
```
```suggestion
          description: "Controls whether the default yank register (`\"`) maps to the system clipboard.",
```

To match how VS Code setting descriptions are phrased.

71 commented on src/state/registers.ts:
```diff
@@ -361,6 +362,10 @@ class ClipboardRegister extends Register implements Register.Readable,
   public readonly flags = Register.Flags.CanRead | Register.Flags.CanWrite;
 
   public async get() {
+    if (!vscode.workspace.getConfiguration(extensionName).get<boolean>("useSystemClipboard")) {
```
Given how often the clipboard might be accessed, I would prefer subscribing to changes to that config value in the `Registers` constructor, similarly to what `Modes` does: https://github.com/71/dance/blob/64bf13e5fab09961952857beae3a2c09f600bff2/src/state/modes.ts#L575-L582

And storing a boolean in `ClipboardRegister`, like `_useSystemClipboard`.

71 reviewed (changes requested):
Thanks! For some reason I thought this was already implemented.

Please change the documentation comment of `ClipboardRegister` to indicate that it will not map to the system clipboard if `useSystemClipboard` is false.

ilyagr commented:
How do you access the clipboard with this setting? Do you rebind Ctrl-V back to paste in insert mode?

I think I'd prefer the default register to be different from the clipboard register. However, that might be more work to implement.

Please feel free to merge this PR without resolving these, since I'm asking for additional features that can be worked on by people who need them in separate PRs.

71 commented:
@ilyagr Yep, similarly to VSCodeVim and others, you'd use the built-in commands to manipulate the clipboard. Further settings to assign another register to the clipboard might be interesting, though.

Even if we only support `null` and `"caret"` (the default) in this PR, it might be interesting to change `"dance.useSystemClipboard": boolean` to `"dance.clipboardRegister": "<register>" | null` for better support of such future changes.

blefevre commented:
Thanks for the feedback, I've adjusted the setting to be a dropdown so the user can select their own clipboard register.
The current options are `"`, `null`, and `A-Z`, but I'm happy to change those as you see fit.

71 commented on src/state/registers.ts:
```diff
@@ -583,6 +585,33 @@ export abstract class RegisterSet implements vscode.Disposable {
+            this.dquote = new ClipboardRegister();
+            this._named.set("dquote", this.dquote);
+          } else if (value != null) {
+            this._named.set(value, new ClipboardRegister());
```
This will lead to strange behaviors if `value` corresponds to a built-in register that is given a property, e.g. ":". Please validate that `value` is a single letter.

blefevre replied:
I've moved the available options for the clipboard register into `meta.ts` so we can check that here to ensure that only valid options are processed.

71 commented on src/state/registers.ts:
```diff
@@ -583,6 +585,33 @@ export abstract class RegisterSet implements vscode.Disposable {
+            this.dquote = new GeneralPurposeRegister('"', "copy");
+            this._named.set("dquote", this.dquote);
+          } else if (this._systemClipbordRegister != null) {
+            this._named.set(this._systemClipbordRegister, new GeneralPurposeRegister(this._systemClipbordRegister, "copy"));
```
We should try to keep the old value of the icon.

71 reviewed (changes requested):
Thanks! I prefer this approach, even though it's a bit less obvious to implement.

ilyagr commented:
> I've adjusted the setting to be a dropdown so the user can select their own clipboard register.

Thank you, I very much appreciate it!

> The current options are ", null, and A-Z, but I'm happy to change those as you see fit.

Vim uses `+` and `*` for X11 selection (which I don't care about). Kakoune doesn't seem to use that. I'm unsure about Dance. If it's easy to make these an option, I think it'd be nice, but picking something between A-Z works too.

blefevre commented:
I think I've addressed all the above feedback including adding `+` and `*`, it should be pretty trivial to add new registers as long as they aren't already in use by Dance.

71 commented on meta.ts:
```diff
@@ -274,6 +274,12 @@ function parseDocComments(code: string, modulePath: string) {
+ * Keys available for binding as the system clipboard register.
+ * These are available in addition to the default `"` and `null`
+ */
+export const availableClipboardRegisters = [..."abcdefghijklmnopqrstuvwxyz".split(""), "*", "+"];
```
```suggestion
export const availableClipboardRegisters = [..."abcdefghijklmnopqrstuvwxyz", "*", "+"];
```

71 replied:
Also, it might be relevant to remove a few letters from the alphabet, as some letters are mapped to special registers (e.g. **d**quote is also mapped to `d` and `D`).

blefevre replied:
> Also, it might be relevant to remove a few letters from the alphabet, as some letters are mapped to special registers (e.g. dquote is also mapped to d and D).
Is this true? I don't see any evidence of this in Dance or Kakoune.

blefevre replied:
Okay I see where this is happening in Dance, but it doesn't seem quite right as half the bindings overwrite the other half, e.g `D` and `d` don't map to `dquote` but `dot`.

I'll remove `acdhspu` from the available registers for the system clipboard but you might want to add explicit definitions for the alternate bindings of special registers.

71 replied:
Welp, that's a great point. I don't remember why it was made this way, but there was definitely an oversight on my part. This is a different issue, so thanks for changing this in the meantime!

71 commented on src/state/registers.ts:
```diff
@@ -583,6 +591,37 @@ export abstract class RegisterSet implements vscode.Disposable {
+        ".systemClipboardRegister",
+        (value) => {
+          if (!['"', null, ...availableClipboardRegisters].includes(value)) {
+            return;
```
I think this should be equivalent to having no system clipboard register.

```suggestion
            value = null;
```

71 replied:
And logging a warning.

71 commented on src/state/registers.ts:
```diff
@@ -583,6 +591,37 @@ export abstract class RegisterSet implements vscode.Disposable {
+
+    // Watch systemClipboardRegister setting and update binding
+    if (extension !== undefined) {
+      extension.observePreference<keyof Registers>(
```
```suggestion
      extension.observePreference<string | null>(
```

71 commented on src/state/registers.ts:
```diff
@@ -583,6 +591,37 @@ export abstract class RegisterSet implements vscode.Disposable {
+          if (this._systemClipboardRegister === "dquote") {
+            this.dquote = new GeneralPurposeRegister('"', "copy");
+            this._named.set("dquote", this.dquote);
+          } else if (this._systemClipboardRegister != null) {
```
```suggestion
          } else if (this._systemClipboardRegister !== undefined) {
```

71 commented on src/state/registers.ts:
```diff
@@ -583,6 +591,37 @@ export abstract class RegisterSet implements vscode.Disposable {
+
+          // Set new value to be a ClipboardRegister
+          if (value === "dquote") {
+            this.dquote = new ClipboardRegister();
```
Perhaps you could rewrite the `dquote` field as a `get` property that reads `_named` to avoid keeping track of both `_named` and `dquote`?

71 reviewed (changes requested):
Thanks! Sorry for the delay, I've been really busy.

I also have a few more comments, but after this we should be good to go.

blefevre commented on src/state/registers.ts:
```diff
@@ -553,7 +561,7 @@ export abstract class RegisterSet implements vscode.Disposable {
   );
 
-  public constructor() {
+  public constructor(extension?: Extension) {
```
`DocumentRegisters` inherits from `RegisterSet` and probably shouldn't be trying to set the system clipboard to a document specific register.

71 replied:
Ah right this is `RegisterSet`, not `Registers`, my bad.

71 commented on src/state/registers.ts:
```diff
@@ -441,8 +441,9 @@ export abstract class RegisterSet implements vscode.Disposable {
-  public dquote: Register = new GeneralPurposeRegister('"', "copy");
-
+  public get dquote(): Register {
+    return this._named.get("dquote") || new GeneralPurposeRegister('"', "copy");
```
Since it's added in the constructor, I think it's fine to expect that it exists.
```suggestion
    return this._named.get("dquote")!;
```

71 commented on src/state/registers.ts:
```diff
@@ -594,30 +595,26 @@ export abstract class RegisterSet implements vscode.Disposable {
           }
 
-          this._systemClipboardRegister = value;
+          this._systemClipboardRegister = value || undefined;
```
Slightly more strict:

```suggestion
          this._systemClipboardRegister = value ?? undefined;
```

71 commented on src/state/registers.ts:
```diff
@@ -594,30 +595,26 @@ export abstract class RegisterSet implements vscode.Disposable {
-            this.dquote = new ClipboardRegister();
-            this._named.set("dquote", this.dquote);
-          } else if (value != null) {
+          if (value != null) {
```
Slightly more strict:
```suggestion
          if (value !== null) {
```

71 reviewed (changes requested):
Some more nitpicks... Hopefully really the last round!

blefevre commented:
Sorry for the delay, this should be good to go now.

71 reviewed (approved):
Thanks!
